### PR TITLE
fix(connector): [Paypal] dispute webhook deserialization failure

### PR DESCRIPTION
**Pull request by KiranKBR**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This refactored branch resolves the PayPal dispute webhook deserialization issue.
-> It was failing to identify the webhook entity type and object reference ID.
-> I updated the fields accordingly to resolve the deserialization issue.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Motivation https://github.com/juspay/hyperswitch/issues/5069

there are some filed refactors included which resolved the problem

## How did you test it?
I ran the ngrok for frowarding webhook payload to localhost with url (https://697b-219-65-110-2.ngrok-free.app/webhooks/merchant_1719300138/paypal)

I have created dispute webhook through paypal simulator and these are the logs :
Incoming webhook Payload:
```
{
  "id": "WH-4M0448861G563140B-9EX36365822141321",
  "create_time": "2018-06-21T13:36:33.000Z",
  "resource_type": "dispute",
  "event_type": "CUSTOMER.DISPUTE.CREATED",
  "summary": "A new dispute opened with Case # PP-000-042-663-135",
  "resource": {
    "disputed_transactions": [
      {
        "seller_transaction_id": "00D10444LD479031K",
        "seller": {
          "merchant_id": "RD465XN5VS364",
          "name": "Test Store"
        },
        "items": [],
        "seller_protection_eligible": true
      }
    ],
    "reason": "MERCHANDISE_OR_SERVICE_NOT_RECEIVED",
    "dispute_channel": "INTERNAL",
    "update_time": "2018-06-21T13:35:44.000Z",
    "create_time": "2018-06-21T13:35:44.000Z",
    "messages": [
      {
        "posted_by": "BUYER",
        "time_posted": "2018-06-21T13:35:52.000Z",
        "content": "qwqwqwq"
      }
    ],
    "links": [
      {
        "href": "https://api.paypal.com/v1/customer/disputes/PP-000-042-663-135",
        "rel": "self",
        "method": "GET"
      },
      {
        "href": "https://api.paypal.com/v1/customer/disputes/PP-000-042-663-135/send-message",
        "rel": "send_message",
        "method": "POST"
      }
    ],
    "dispute_amount": {
      "currency_code": "USD",
      "value": "3.00"
    },
    "dispute_id": "PP-000-042-663-135",
    "dispute_life_cycle_stage": "INQUIRY",
    "status": "OPEN"
  },
  "links": [
    {
      "href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-4M0448861G563140B-9EX36365822141321",
      "rel": "self",
      "method": "GET",
      "encType": "application/json"
    },
    {
      "href": "https://api.paypal.com/v1/notifications/webhooks-events/WH-4M0448861G563140B-9EX36365822141321/resend",
      "rel": "resend",
      "method": "POST",
      "encType": "application/json"
    }
  ],
  "event_version": "1.0"
}
```

Dispute Creation:
image

Source Verification:
Due to these are Mock events , we cant verify them through Paypal so I have done source verification as true to check further flows like Database record creation and triggering outgoing webhooks.

Database Record Creation:
image

Outgoing Webhook Response Payload:
```
{
    "merchant_id": "merchant_1719481316",
    "event_id": "evt_0190594e16ab7f38abd50ca6f32707b9",
    "event_type": "dispute_opened",
    "content": {
        "type": "dispute_details",
        "object": {
            "dispute_id": "dp_rgC7BN7DQRQ9YlksqCmO",
            "payment_id": "pay_qW4qVbLZXOO1GQsd0mQ3",
            "attempt_id": "00D10444LD479031K",
            "amount": "300",
            "currency": "USD",
            "dispute_stage": "pre_dispute",
            "dispute_status": "dispute_opened",
            "connector": "paypal",
            "connector_status": "Open",
            "connector_dispute_id": "PP-000-042-663-135",
            "connector_reason": "MERCHANDISE_OR_SERVICE_NOT_RECEIVED",
            "connector_reason_code": "MERCHANDISE_OR_SERVICE_NOT_RECEIVED",
            "challenge_required_by": null,
            "connector_created_at": "2018-06-21T13:35:44.000Z",
            "connector_updated_at": "2018-06-21T13:35:44.000Z",
            "created_at": "2024-06-27T10:46:48.214Z",
            "profile_id": "pro_AKqLxMihwY7h8QqvuPfH",
            "merchant_connector_id": "mca_q1lwLDnNEPzyGvClsHsR"
        }
    },
    "timestamp": "2024-06-27T10:46:48.235Z"
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Review by swangi-kumari — APPROVED** [review]

**Review by srujanchikke — DISMISSED** [review]

File: crates/router/src/connector/paypal.rs [hunk 1]
```diff
@@ -1534,9 +1533,9 @@ impl api::IncomingWebhook for Paypal {
                 Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
                     api_models::payments::PaymentIdType::PaymentAttemptId(
                         resource
-                            .dispute_transactions
+                            .disputed_transactions
                             .first()
-                            .map(|transaction| transaction.reference_id.clone())
+                            .map(|transaction| transaction.seller_transaction_id.clone())
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/paypal.rs:1538]: If `seller_transaction_id` is not our id, we should change PaymentIdType to ConnectorTransactionId instead of PaymentAttemptId

File: crates/router/src/connector/paypal.rs [hunk 2]
```diff
@@ -1553,17 +1552,17 @@ impl api::IncomingWebhook for Paypal {
             .parse_struct("PaypalWebooksEventType")
             .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
         let outcome = match payload.event_type {
-            PaypalWebhookEventType::CustomerDisputeCreated
-            | PaypalWebhookEventType::CustomerDisputeResolved
-            | PaypalWebhookEventType::CustomerDisputedUpdated
-            | PaypalWebhookEventType::RiskDisputeCreated => Some(
+            PaypalWebhookEventType::CustomerDisputeResolved => Some(
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/paypal.rs:1555]: We should check the outcome code for all the Dispute events right, why is this change required here?

File: crates/router/src/connector/paypal.rs [hunk 3]
```diff
@@ -1608,27 +1607,36 @@ impl api::IncomingWebhook for Paypal {
...
-            connector_reason_code: payload.external_reason_code,
-            challenge_required_by: payload.seller_response_due_date,
-            created_at: payload.create_time,
-            updated_at: payload.update_time,
-        })
+        match webhook_payload.resource {
+            transformers::PaypalResource::PaypalCardWebhooks(_)
+            | transformers::PaypalResource::PaypalRedirectsWebhooks(_)
+            | transformers::PaypalResource::PaypalRefundWebhooks(_) => {
+                Err(errors::ConnectorError::WebhookResourceObjectNotFound.into())
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/paypal.rs:1618]: Instead of throwing Badrequest (4xx) here, can we make it as a 5xx?

File: crates/router/src/connector/paypal.rs (hunk 2 above)
Comment (KiranKBR) [reply crates/router/src/connector/paypal.rs:1555]: They r sending outcome_code for resolved events only

File: crates/router/src/connector/paypal.rs (hunk 1 above)
Comment (KiranKBR) [reply crates/router/src/connector/paypal.rs:1538]: This should be seller_transaction_id

File: crates/router/src/connector/paypal.rs [hunk 4]
```diff
@@ -1622,27 +1621,36 @@ impl api::IncomingWebhook for Paypal {
...
-            connector_reason_code: payload.external_reason_code,
-            challenge_required_by: payload.seller_response_due_date,
-            created_at: payload.create_time,
-            updated_at: payload.update_time,
-        })
+        match webhook_payload.resource {
+            transformers::PaypalResource::PaypalCardWebhooks(_)
+            | transformers::PaypalResource::PaypalRedirectsWebhooks(_)
+            | transformers::PaypalResource::PaypalRefundWebhooks(_) => {
+                Err(errors::ConnectorError::ResponseDeserializationFailed.into())
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/paypal.rs:1632]: Can we have an attach_printable here?

File: crates/router/src/connector/paypal.rs [hunk 5]
```diff
@@ -1567,17 +1566,17 @@ impl api::IncomingWebhook for Paypal {
...
                 request
                     .body
                     .parse_struct::<paypal::DisputeOutcome>("PaypalWebooksEventType")
                     .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?
                     .outcome_code,
             ),
-            PaypalWebhookEventType::PaymentAuthorizationCreated
+            PaypalWebhookEventType::CustomerDisputeCreated
+            | PaypalWebhookEventType::RiskDisputeCreated
+            | PaypalWebhookEventType::CustomerDisputedUpdated
```
[the remark this answers is not in the document]
Comment (KiranKBR) [reply crates/router/src/connector/paypal.rs:1578]: Bro, I have already told  we r not getting outcome_code

File: crates/router/src/connector/paypal.rs (hunk 4 above)
Comment (KiranKBR) [reply crates/router/src/connector/paypal.rs:1632]: we cant do for ResponseDeserializationFailed

File: crates/router/src/connector/paypal.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sai-harsha-vardhan) [reply crates/router/src/connector/paypal.rs:1578]: Can you please share the sample payload for CustomerDisputedUpdated event?

[the remark this answers is not in the document]
Comment (KiranKBR) [reply crates/router/src/connector/paypal.rs:1578]: I have shared u via slack

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by swangi-kumari — APPROVED** [review]
